### ISSUE TITLE
Support Apache Pulsar message broker in Celery 4.x / 5.0

### ISSUE THREAD

**Issue by madprogrammer** [celery/celery#5487]
# Checklist

- [X] I have checked the [issues list](https://github.com/celery/celery/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22Issue+Type%3A+Feature+Request%22+)
  for similar or identical feature requests.
- [X] I have checked the [pull requests list](https://github.com/celery/celery/pulls?utf8=%E2%9C%93&q=is%3Apr+label%3A%22PR+Type%3A+Feature%22+)
  for existing proposed implementations of this feature.
- [X] I have checked the [commit log](https://github.com/celery/celery/commits/master)
  to find out if the if the same feature was already implemented in the
  master branch.
- [X] I have included all related issues and possible duplicate issues
  in this issue (If there are none, check this box anyway).

#### Related Issues

- None

#### Possible Duplicates

- None

# Brief Summary

[Apache Pulsar](https://pulsar.apache.org/en/) is a relatively new, but powerful and highly scalable messaging system, offering feature set similar to RabbitMQ and Kafka. Official client library exists for Python.

# Design

## Architectural Considerations

Add Pulsar backend to Celery. `Shared` consumer mode allows multiple consumers to subscribe to the same topic (queue) and get messages in a round-robin fashion, with support for consumer-level (but not message-level) priority.

## Proposed Behavior

Celery is able to use Pulsar as underlying message broker, similar to RabbitMQ.

## Diagrams

N/A

## Alternatives

None

**Comment by auvipy** [comment]
might be moved in 5.1 later

**Comment by auvipy** [comment]
but need funding and strong community demand

**Comment by truong-hua** [comment]
Pulsar is a great persistent Message Queue and scalable regarding disk. Directly support Pulsar will help Celery can queue a large number of tasks in backlog without performance degrading as in RabbitMQ.

Besides, Pulsar support message replay by just modifying the subscription cursor which will help replay/retry history tasks much more easier even they were executed years ago.

Thirdly, Pulsar can directly receive message event from any producer that supports Kafka/Pulsar/AMQP/HTTP protocol which helps Celery can directly integrate with may be everything in the MQ community without complex routing between Pulsar and RabbitMQ.

We are using both Pulsar and RabbitMQ and the maintenance cost for RabbitMQ is much higher because you have to drop message in large queue or make sure that the consuming speed is capable to keep queue size small. Otherwise, there will be a big trouble with RabbitMQ. Nowadays RabbitMQ has the Lazy queue but it is still can not compete with Pulsar regarding persistent queue scalability.

This feature is not just about supporting more broker but if we can consume the more flexible message format. Celery can be consider or used to integrate with any event like producer.

**Comment by adpande** [comment]
> @auvipy we are experimenting https://streamnative.io/blog/tech/2020-06-15-announcing-aop-on-pulsar/
> 
> I will let you know if it works properly. If so, I think we just need to update our document to mention about the version of AMQP protocol supported as broker rather than directly mention about RabbitMQ and only RabbitMQ currently.
> 
> There is also https://github.com/strimzi/strimzi-kafka-bridge for Kafka backed cluster.
> 
> I think we just need to make sure the AMQP compatibility in every Celery version. I will save a lot of time. I guess so.

Hey, did it work?

**Comment by truong-hua** [comment]
No, @adpande , Celery is using some structured meta data of a message which is not fully supported by the aop. We modified the aop to make it compatible with Celery too. After that, it's working but the aop has a lot of limitation that sacrifice the scalability of Pulsar.

If you really need the Pulsar, I recommend the structure that Pulsar --- (Pulsar function to forward message to RabbitMQ) ---> Celery. So you will need all Pulsar, RabbitMQ.

**Comment by adpande** [comment]
Thanks @truong-hua for the update. Keeping 2 msg brokers in overkill for my problem. Lets see how it goes.

**Comment by asrocha** [comment]
Hi!
@truong-hua  and @adpande  work fine ? use rabbitMQ as pulsar Sink ? I have the same idea, but not time to test.

**Comment by truong-hua** [comment]
> Hi! @truong-hua and @adpande work fine ? use rabbitMQ as pulsar Sink ? I have the same idea, but not time to test.

Yes, it's working and we are using it in production environment for a while. But you will have to extends the ConsumerStep to create your owned because the message published from Pulsar to RabbitMQ does not have the format and headers Celery need. FYI https://docs.celeryq.dev/en/stable/userguide/extending.html. Or you can rewrite your RabbitMQ Sink for Pulsar that support Celery but I don't think it's a good idea.

**Comment by asrocha** [comment]
@truong-hua 

Thanks, I will work in this solution and feedback here

Thanks so much !

**Comment by auvipy** [comment]
we will first need to add it to kombu

**Comment by TechWithTy** [comment]
Ant updates?
